### DISCUSSION TITLE
Build Unit Test failures after a076b9d7 ("Audit all transitive dependencies and fix latest vulnerabilities", 2024-12-27)

### DISCUSSION THREAD

**Discussion by bevanweiss** [wixtoolset/issues#8884] [WiX Development]
@robmen / @barnson 
Is anyone else seeing unit test failures when doing `.\devbuild.cmd` after commit a076b9d7?

```
[xUnit.net 00:00:02.93]     DutilTests.ApupUtil.AllocChainFromAtomSortsDescending [FAIL]
  Failed DutilTests.ApupUtil.AllocChainFromAtomSortsDescending [439 ms]
  Error Message:
   WixInternal.TestSupport.SucceededException : WixAssert.Succeeded() Failure
HRESULT:  0x80040154
Message:  Failed to parse feed: ...\wix\build\libs\Debug\x86\TestData\ApupUtilTests\FeedBv2.0.xml
  Stack Trace:
     at WixInternal.TestSupport.WixAssert.Succeeded(Int32 hr, String format, Object[] formatArgs) in ...\wix\src\internal\WixInternal.TestSupport\XunitExtensions\WixAssert.cs:line 85
   at WixInternal.TestSupport.NativeAssert.Succeeded(Int32 hr, SByte* zFormat, Char*[] wzArgs) in ...\wix\src\internal\WixInternal.TestSupport.Native\NativeAssert.h:line 64
   at DutilTests.ApupUtil.AllocChainFromAtomSortsDescending() in ...\wix\src\libs\dutil\test\DUtilUnitTest\ApupUtilTests.cpp:line 30
[xUnit.net 00:00:02.46]     DutilTests.MonUtil.MonUtilTest [SKIP]
[xUnit.net 00:00:02.47]     DutilTests.FileUtil.FileUtilTest [SKIP]
  Skipped DutilTests.MonUtil.MonUtilTest [1 ms]
  Skipped DutilTests.FileUtil.FileUtilTest [1 ms]
[xUnit.net 00:00:03.12]     DutilTests.LocUtil.CanLoadStringsWxl [FAIL]
  Failed DutilTests.LocUtil.CanLoadStringsWxl [455 ms]
  Error Message:
   WixInternal.TestSupport.SucceededException : WixAssert.Succeeded() Failure
HRESULT:  0x80040154
Message:  Failed to parse strings.wxl: ...\wix\build\libs\Debug\x64\TestData\LocUtilTests\strings.wxl
  Stack Trace:
     at WixInternal.TestSupport.WixAssert.Succeeded(Int32 hr, String format, Object[] formatArgs) in ...\wix\src\internal\WixInternal.TestSupport\XunitExtensions\WixAssert.cs:line 87
   at DutilTests.LocUtil.CanLoadStringsWxl() in ...\wix\src\libs\dutil\test\DUtilUnitTest\LocUtilTests.cpp:line 33
Results File: ...\wix\src\libs\..\..\build\logs\TestResults\libs.trx

Failed!  - Failed:     1, Passed:   102, Skipped:     2, Total:   105, Duration: 2 s - DUtilUnitTest.dll (netcoreapp1.0)
Failed!  - Failed:     1, Passed:   102, Skipped:     2, Total:   105, Duration: 2 s - DUtilUnitTest.dll (netcoreapp1.0)
```

**Comment by bevanweiss** [comment]
It looks like reverting the following changes resolves the test breaks (at least for me).

```diff
src/internal/WixInternal.TestSupport/WixInternal.TestSupport.csproj
index 4e1c3c26..1b713a0d 100644
@@ -4,7 +4,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0;net472</TargetFrameworks>
+    <TargetFrameworks>net6.0;net472</TargetFrameworks>
     <IsPackable>true</IsPackable>
     <DebugType>embedded</DebugType>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
```

```diff
src/wix/WixInternal.Core.TestPackage/WixInternal.Core.TestPackage.csproj
index 9d242b6c..99ffe620 100644
@@ -3,7 +3,7 @@
 
 <Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>netstandard2.0</TargetFrameworks>
+    <TargetFrameworks>net6.0;net472</TargetFrameworks>
     <Description>Internal WiX Toolset Test Package</Description>
     <DebugType>embedded</DebugType>
     <PublishRepositoryUrl>true</PublishRepositoryUrl>
```

I think if we could get away with it, just having netstandard2.0 as the TargetFrameworks for these libraries should 'be fine'.
netstandard2.0 is supported on net461+ (although net472+ is recommended by MS)

**Reply by robmen** [reply]
Hmm, I'm not seeing that test failure, and more importantly, the build server isn't seeing it either. However, I tested reverting the `net6.0` back to `netstandard2.0` in `WixInternal.TestSupport.csproj`, which also seems to work. The `net472` was still required for me to build native tests.

I have a PR (https://github.com/wixtoolset/wix/pull/585) out with that change and a few small cleanup bits. I hope that helps address your local build issues.

**Reply by bevanweiss** [reply]
> Hmm, I'm not seeing that test failure, and more importantly, the build server isn't seeing it either. However, I tested reverting the `net6.0` back to `netstandard2.0` in `WixInternal.TestSupport.csproj`, which also seems to work. The `net472` was still required for me to build native tests.
> 
> I have a PR ([wixtoolset/wix#585](https://github.com/wixtoolset/wix/pull/585)) out with that change and a few small cleanup bits. I hope that helps address your local build issues.

Still one failure remaining.
```
[xUnit.net 00:00:02.17]     DutilTests.LocUtil.CanLoadStringsWxl [FAIL]
  Failed DutilTests.LocUtil.CanLoadStringsWxl [420 ms]
  Error Message:
   WixInternal.TestSupport.SucceededException : WixAssert.Succeeded() Failure
HRESULT:  0x80040154
Message:  Failed to parse strings.wxl: ...\wix\build\libs\Debug\x86\TestData\LocUtilTests\strings.wxl
  Stack Trace:
     at WixInternal.TestSupport.WixAssert.Succeeded(Int32 hr, String format, Object[] formatArgs) in ...\wix\src\internal\WixInternal.TestSupport\XunitExtensions\WixAssert.cs:line 8     at WixInternal.TestSupport.WixAssert.Succeeded(Int32 hr, String format, Object[] formatArgs) in ...\wix\src\internal\WixInternal.TestSupport\XunitExtensions\WixAssert.cs:line 85
   at WixInternal.TestSupport.NativeAssert.Succeeded(Int32 hr, SByte* zFormat, Char*[] wzArgs) in ...\wix\src\internal\WixInternal.TestSupport.Native\NativeAssert.h:line 64
   at DutilTests.LocUtil.CanLoadStringsWxl() in ...\wix\src\libs\dutil\test\DUtilUnitTest\LocUtilTests.cpp:line 31
```
If I change things like net472=>net46, or net6.0=>netstandard2.0 then I can sometimes get a slightly different DUtil UnitText XML failure.  Which perhaps leads to it being either XML related, or some concurrency issue with DUtil (if it's not just a bug somewhere in my build environment).

The build environment is;
Microsoft Windows [Version 10.0.27764.1000] (insider canary channel, so it's possibly this... but older commits work)
Microsoft Visual Studio Enterprise 2022 (64-bit) - Current (Version 17.12.3)
but I do also have Visual Studio Preview 2022 installed (maybe there is some cross contamination.. but again, older commits work)

**Reply by barnson** [reply]
I've intermittently gotten failures like that but they always go away on a rebuild. No previews.

**Reply by borrrden** [reply]
I get this failure with strings.wxl and controls.wxl every time I try to build with the latest vs2022 on windows 10.0.19045.0.  Note that the HRESULT there corresponds to `REGDB_E_CLASSNOTREG`

**Reply by robmen** [reply]
Yes. It's become more of a problem lately. No one has tracked down the root cause yet.

**Reply by Nico-Ko** [reply]
I investigated the problem a little bit and it seems to be a timing Problem, but i do not know where the problem comes from.

I added a retry after 5s to the XmlLoadDocumentFromFile function from xmlutil.cpp like this:
```
extern "C" HRESULT DAPI XmlLoadDocumentFromFile(
    __in_z LPCWSTR wzPath,
    __out IXMLDOMDocument** ppixdDocument
)
{
    HRESULT hr = XmlLoadDocumentFromFileEx(wzPath, 0, ppixdDocument);

    if (FAILED(hr))
    {
        Sleep(5000);
        hr = XmlLoadDocumentFromFileEx(wzPath, 0, ppixdDocument);
    }
    return hr;
}
```

The problem did not occur anymore and in the test execution time it is visible that the retry is used.
![image](https://github.com/user-attachments/assets/d847569b-7a3e-4bd8-882c-fe6563b32099)

The problem seems to be fixable with way smaller delays (5ms) I just used 5s for better visibility in the test results.

**Reply by robmen** [reply]
@Nico-Ko Thanks for the retry suggestion. I've added it in wixtoolset/wix#642, so hopefully, that will resolve these test case intermittent failures for good.
